test/test_add.cc

```diff
@@ -205,38 +71,16 @@ void test_add_work(Params& params, bool run)
         {slate::Option::Target, target}
     };
 
-    // MPI variables
-    int mpi_rank, myrow, mycol;
-    MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
-    gridinfo(mpi_rank, p, q, &myrow, &mycol);
-
-    // Matrix A: figure out local size.
-    int64_t mlocA = num_local_rows_cols(m, nb, myrow, p);
-    int64_t nlocA = num_local_rows_cols(n, nb, mycol, q);
-    int64_t lldA  = blas::max(1, mlocA); // local leading dimension of A
-    std::vector<scalar_t> A_data, B_data;
-
-    int64_t nlocB = nlocA, lldB = lldA;
-
-    slate::Matrix<scalar_t> Afull, Bfull;
-    if (origin != slate::Origin::ScaLAPACK) {
-        // SLATE allocates CPU or GPU tiles.
-        slate::Target origin_target = origin2target( origin );
-        Afull = slate::Matrix<scalar_t>( m, n, nb, p, q, MPI_COMM_WORLD);
-        Afull.insertLocalTiles( origin_target );
-        Bfull = slate::Matrix<scalar_t>( m, n, nb, p, q, MPI_COMM_WORLD);
-        Bfull.insertLocalTiles( origin_target );
-    }
-    else {
-        // Allocate ScaLAPACK data.
-        A_data.resize( lldA*nlocA );
-        B_data.resize( lldB*nlocB );
-        // Create SLATE matrix from the ScaLAPACK layout.
-        Afull = slate::Matrix<scalar_t>::fromScaLAPACK(
-                    m, n, &A_data[0], lldA, nb, p, q, MPI_COMM_WORLD );
-        Bfull = slate::Matrix<scalar_t>::fromScaLAPACK(
-                    m, n, &B_data[0], lldB, nb, p, q, MPI_COMM_WORLD );
-    }
+    auto A_alloc = allocate_test_Matrix<scalar_t>( check || ref, false, m, n, params );
+    auto B_alloc = allocate_test_Matrix<scalar_t>( check || ref, false, m, n, params );
+
+    auto& Afull     = A_alloc.A;
+    auto& Bfull     = B_alloc.A;
+    auto& Aref_full = A_alloc.Aref;
+    auto& Bref_full = B_alloc.Aref;
+    auto& Aref_data = A_alloc.Aref_data;
+    auto& Bref_data = B_alloc.Aref_data;
```

Comment:
That's probably a good idea.  In addition to the strangeness, `matrix` is just more obvious.  Although, since this will be a bigger refactor, it might be better to save it until we're working on #157.

Reply:
Thanks for pointing out #157. Yes, we can consider some changes after getting both of these merged.